### ISSUE TITLE
enforce_kl=true silently disables the watermark with bf16 logits (NaN in kl_from_delta collapses the bisection to delta=0)

### ISSUE THREAD

**Issue by Kx-Feng** [eth-sri/diffusion-lm-watermark#1]
### Summary

With `enforce_kl: true` (the ε-parameterization), the watermark is **silently disabled** whenever the model produces **bfloat16 logits** (e.g., any bitsandbytes-quantized or bf16 deployment). Generation completes normally and the text is coherent, but the detector finds a green fraction ≈ γ, i.e. no watermark at all. No error or warning is raised, which makes this easy to miss — `enforce_kl: false` (δ-parameterization) works fine under the same setup, so everything looks healthy unless one specifically checks detection under the KL mode.

### Environment

- `GSAI-ML/LLaDA-8B-Base` loaded in 8-bit via bitsandbytes (model logits are bf16); torch 2.10, transformers 4.57
- Config: main Llada config with `enforce_kl: true`, temperature 0.5, `C={-1}`, sumhash, bernoulli γ=0.25
- The issue is dtype-generic: it should reproduce in any pipeline where the logits reaching `compute_logit_booster` are bf16 (and can also trigger in fp32 for very sharp distributions, see below).

### Symptom

| setting | green fraction of generated text | binomial p-value |
|---|---|---|
| `enforce_kl: false`, δ=4 | ≈ 0.55 | ~1e-8 |
| `enforce_kl: true`, δ (KL budget) = 0.25 or 0.5 | **≈ 0.19–0.24 (≈ γ)** | 0.6–0.9 |

Text quality in both cases is normal — the KL-mode runs simply contain no watermark.

### Root cause

In `OurWatermark.kl_from_delta` (src/dlm_watermark/watermarks/diffusion_watermark.py):

```python
kl = torch.sum(q * torch.log(q / probs), dim=-1, keepdim=True)
kl[kl.isnan()] = float("inf")  # Avoid NaNs in the KL divergence
```

With bf16 logits, `softmax(logits / T)` underflows to **exact zeros** for tail tokens (bf16 loses to zero around e⁻⁸⁸; at T=0.5 a logit spread of ~45 is enough). Both `q` and `probs` then contain exact zeros at the same tail positions, and

```
q * log(q / probs)  →  0 * log(0/0)  →  0 * NaN  →  NaN
```

so `kl` is NaN for essentially **every position at every bisection iteration**. The NaN→inf line then marks each position as "over budget", `find_delta`'s bisection shrinks `hi` toward `lo = 0` for 16 iterations, and the returned per-position δ is ~0. The logit booster becomes ~0 and the watermark is a no-op.

### Suggested fix

Compute the KL in float32 and use clamped logs so that underflowed tail terms become negligible instead of NaN, while a genuine budget violation (large q mass where p underflowed) still yields a huge finite KL that the bisection correctly treats as over budget:

```python
q = F.softmax((logits.float() + logit_booster) / self.temperature, dim=-1)
p = probs.float()

kl_terms = torch.where(
    q > 0,
    q * (q.clamp_min(1e-45).log() - p.clamp_min(1e-45).log()),
    torch.zeros_like(q),
)
kl = kl_terms.sum(dim=-1, keepdim=True).clamp(min=0.0)
```

One subtlety we hit while fixing this: special-casing `q > 0 & p == 0` as `inf` is *not* a valid alternative — bf16 `p` underflows around e⁻⁸⁸ while f32 `q` survives to ~e⁻¹⁰³, so that band gets flagged spuriously and the bisection collapses again. The clamped-log formulation handles both regimes gracefully (an underflowed term contributes `q·(log q + 103) ≈ 0` for tiny q, but ≈ 50 nats for macroscopic q).

### Verification after the fix

Probing `watermark_logits` on real LLaDA-8B bf16 logits (prompt + 150 masked positions, first denoising step):

```
enforce_kl budget 0.5:  per-position KL pinned at 0.500, standardized boost shift ≈ 0.94σ
enforce_kl budget 1.0:  per-position KL pinned at 1.000, shift ≈ 1.28σ
(reference, enforce_kl=false δ=4:  mean KL 0.27, shift ≈ 0.72σ)
```

and end-to-end generation recovers the expected green-fraction lift.

### Possibly related minor issue

While debugging we also noticed that the "skip the prompt" optimization in `compute_logit_booster` reads `masked_inputs[:, 0]` (the first *column*, i.e. position 0 across the batch) rather than the first masked index per row, so `start_idx` is effectively always 0 whenever position 0 is unmasked. Harmless (just wasted compute on prompt positions), but probably not what was intended.

Happy to open a PR with the patch if useful. Thanks for releasing the code — the repo reproduces the paper's δ-parameterization results very cleanly (we measured TPR@1%FPR ≈ 96–100% at δ=4, 150 tokens, on an 8-bit quantized LLaDA-8B-Base).
